README.md

```diff
@@ -1 +1,279 @@
+sort keys (one or more of primary, second, etc key).  To deal with
+memory limits, whenever this table hits a configured size limit (related
+to the amount of memory dedicated to the worker),
+it is simply spilled to disk by writing the entries to disk
+in a key-sorted order.  At end-of-stream, the tables are all opened
+and read in parallel and any entry that is common across the streams is
+aggregated by the aggregation operator and the results are streamed
+to the downstream entity.
+
+This approach can be parallelized across N nodes where the time range of the
+traversal is divided amongst the different nodes and the final step merges
+the streamed results from all of the nodes.
+
+That will work fine for many use cases.
+For many aggregations, the group-by table will stay small when the cardinality
+of the key space is low.
+If you have a feel for how much memory you need ahead of time, you could
+launch N jobs in parallel so a high-cardinality group-by table
+gets split out efficiently.  In this case, the key space isn't disjoint
+but it might not matter much if the sub-cardinality of each key is small
+(meaning there is little advantage to a shuffle)
+so in general this could work well depending on the underlying data
+characteristic.  And if the memory limit is hit here, the table is spilled
+to disk as above.
+
+### version 2: a dynamic model
+
+There are two key problems with the static approach.  First, you don't generally
+know how to choose N so it's unreasonable to ask the user to specify this.
+And second, partitioning the work by time means the key space is replicated
+over each node making the tables inefficient and larger than they need to be.
+
+A solution to these problems is, firstly, to fork a new worker only after the
+aggregation table gets too big (thus providing scale-out memory along with
+scale-out throughput) and, secondly, to implement a shuffle to
+route the table rows to the worker that owns the row.  Ownership of rows
+is handled like spark and map-reduce using a hash partition.  The
+number of dynamic nodes is limited either because the cluster
+manager (e.g., kubernetes) indicates no more resources are available
+or a configured scale-out limit is hit.  Either way, when there are no
+more available resources, workers then spill to disk.
+
+In this approach, we begin with N=M workers for throughput parallelism then
+dynamically add workers (increasing N) to provide additional memory scale-out
+and throughput parallelism.  The traversal range is divided among the M workers
+in some way (i.e., evenly divided, partitioned with a block/M stride, etc)
+
+When a worker hits a memory limit because its aggregation table gets to big,
+it splits its job, say in 2 (though this fanout can be configured), or it resorts
+to spilling if no resource is available for the split.  When it splits,
+half of the key space and half of the input partition goes to the new child.
```

Comment:
It would be good to clarify if/how the keyspace is partitioned across the initial M workers.